### ISSUE TITLE
What to make nullable and what not?

### ISSUE THREAD

**Issue by pauldendulk** [Mapsui/Mapsui#1360]
### How to go about nullability?

With the nullability checks added we are forced to think more about nullability. How I look at it now:
- It is better for things to be non-nullable because consuming classes don't need to check for null.
- For non-nullable you need to either 
  - Initialize the field.
  - Initialize the property.
  - Initialize from constructor arguments.
- Setting some default value seems a simple solution but if you do so you will need a default value that makes sense and often there is no such default. For instance, if a feature has no geometry what is the default Extent? An Extent of 0,0,0,0 is problematic. Suppose you have a layer with features in the range 100,100,200,200. If you add one empty feature with 0,0,0,0 the extent would become 0,0,200,200. So my conclusion is that Feature.Extent should either be nullable or all feature should have a constructor that enforces that the geometry is passed in and a realistic Extent is set. I now propose to do the latter.
- With init properties you still need a default value because the init property does not have to be set by the user. If there is no sensible default you will still need a constructor.
- Some fields have no sensible default and will have to be null.

### About some specific fields

- IFeature.Extent: **not** nullable. Set in constructor.
- IFeature.Geometry/Rect/Point/Raster: **not** nullable. Set in constructor.
- IProvider.GetExtents(): nullable. Unfortunately there could be no features only a null value would make sense.
- ILayer.Extent: nullable. Because derived from provider.
- IMap.Extent: nullable. Because derived from layers.
- IViewport.Envelope: Not sure. I think nullable. You could choose some default but if this is based on non existing data they may result in weird behavior. A zero Width Envelope could cause problems (like divide by zero). Some random value could cause fetching of irrelevant data. Note, this problem may only be relevant during startup while data is loaded but this is an important and often problematic part which we need to get right.
- INavigator.NavigateTo(MRect). Not sure. If it does not allow null the user will need to check. But if the user navigates to a null value (and nothing) happens this may also cause hard to track problems. I now think it should be not-nullable. 

There is a lot more to think about but these were just a few things I encounted.

**Comment by pauldendulk** [comment]
This comment describes my current ideas about which fields should be nullable (or not). I intend to update this comment as my ideas change or expand.

### Nullability guidelines


| Kind | Nullability | Remark |
|---|---|---|
| IFeature.Extent | not-null | Set in constructor. |
| IFeature.Geometry/Rect/etc | not- null | Set in constructor. |
| IProvider.GetExtents() | null | Unfortunately there could be no features only a null value would make sense. |
| ILayer.Extent | null | Because derived from provider. |
| IMap.Extent | null | Because derived from layers. |
| IViewport.Envelope | null | Not sure. I think nullable. You could choose some default but if this is based on non existing data they may result in weird behavior. A zero Width Envelope could cause problems (like divide by zero). Some random value could cause fetching of irrelevant data. Note, this problem may only be relevant during startup while data is loaded but this is an important and often problematic part which we need to get right. |
| INavigator.NavigateTo(MRect) | not-null | Not sure. If it does not allow null the user will need to check. But if the user navigates to a null value (and nothing) happens this may also cause hard to track problems. I now think it should be not-nullable.  |
| Styles | not-null | In many cases a default of Color.Transparent or 0 would make sense.

**Comment by pauldendulk** [comment]
Above IFeature.Extent is suggested to be not-null. For that IFeature.Geometry needs to be not-null but also not Empty. A geometry can be empty if a polygon contains no vertices. We could say this is corrupt data. A Geometry is not allow to be empty within a feature, but that is hard to enforce. If a user edits a polygon manually in say the Mapsui.Samples.Wpf.Editing sample, this is hard to prevent, although possible. Perhaps the editor can delete te feature once the last vertex is removed. Not sure yet, still undecided about this.

**Comment by pauldendulk** [comment]
One general thing. If there is a list of things then it is always an option to send an empty list. I think this should be preferred.

**Comment by pauldendulk** [comment]
@inforithmics Went through the null checks in Mapsui.Core: https://github.com/Mapsui/Mapsui/pull/1392

- There is one in AnimatedFeatures that is problematic and needs a design change.
- There are a lot in GradientTheme I did not look at all.
- All the others are fixed.

My choice is now to let IMap.CRS and IProvider.CRS be null. If you have no idea what it should be then null is the most explicit way to indicate this.

With respect to styles I do not know what the best design would be. Also I did not spend much time on it yet.

If you have any thoughts, suggestions or doubts please comment here. It is difficult to make good choices for everything.

**Comment by pauldendulk** [comment]
Another general guideline. If you pass in an argument to a method that can often be non-null. The caller could check for that. If it si a stateful object it is often harder to enforce non-null.

**Comment by inforithmics** [comment]
Yes I saw the problem with the Styles too, I have to think about it. Maybe (Extension) Methods for every property like ResolvedBackgroundColor that don't return nullable values but good defaults could help. But I have to explore this what's a good solution here.

**Comment by pauldendulk** [comment]
I have been working on editing geometries and it looks like it is the best option to make Geometries nullable and as a consequence IFeature.Extent will be nullable. Not changing it in the table above right now. I would like to revisit is before making a definitive decision.
